kame/kame/pvcsif/pvcbridge.8
.Dd June 25, 1999
.Dt PVCBRIDGE 8
.Os BSD
.Sh NAME
.Nm pvcbridge
.Nd display or control ATM PVC bridging between two PVC sub interfaces
.Sh SYNOPSIS
.Nm pvcbridge
.Op get
.Ar interface
.Nm pvcbridge
.Li add
.Ar interface1
.Ar interface2
.Nm pvcbridge
.Li delete
.Ar interface1
.Ar interface2
.Sh DESCRIPTION
.Nm 
displays or controls bridging between two PVC sub interfaces.
ATM bridging allows to forward ATM frames at the driver level, 
neither IP address nor IP routing is rquired for the PVC sub
interfaces.
An ATM VC should be assgined to each PVC sub interface by 
.Xr pvctxctl 8 .
.Pp
With keyword
.Dq Li get ,
.Nm 
displays the current bridge setting of the given PVC sub interface.
The
.Ar inerface
parameter must be an ATM PVC sub interface created by
.Xr pvcsif 8 .
With keyword
.Dq Li add
or
.Dq Li delete ,
.Nm
sets or deletes bridging between the given two ATM PVC sub interfaces.
With no keyword,
.Dq Li get
is assumed.
.\"
.Sh EXAMPLE
For example, to set up bridging between VC:201 on en0 and VC:202 on en1:
.Pp
.Bd -literal
# pvcsif en0      		# creates pvc0
# ifconfig pvc0 up		# up pvc0
# pvctxctl pvc0 201 -b 45M	# assign vc
# pvcsif en1      		# creates pvc1
# ifconfig pvc1 up		# up pvc1
# pvctxctl pvc1 202 -b 45M	# assign vc
# pvcbridge pvc0 pvc1		# set bridging
.Ed
.\"
.Sh SEE ALSO
.Xr ifconfig 8 , 
.Xr pvcsif 8 ,
.Xr pvctxctl 8


CHANGELOG
$KAME: CHANGELOG,v 1.2458 2003/08/18 05:39:09 itojun Exp $

<200308>
Sun Aug 17 JST 2003  itojun@iijlab.net
	* sys/netinet{,6}/in{,6}_proto.c: enforce inbound IPsec policy on
	  protocols that falls into raw socket as a wildcard case.

2003-08-15  SUZUKI, Shinsuke <suz@crl.hitachi.co.jp>
	* kame/kame/pim6sd/{cfparse.y, mrt.[ch], pim6_proto.c, rp.c}
	Now static-RP configuration can work together with bootstrap-
	message-based RP configuration

2003-08-10  SUZUKI, Shinsuke <suz@crl.hitachi.co.jp>
	* kame/{freebsd5, kame}
	sync with FreeBSD-5.1 RELEASE (behaviour is not confirmed yet)

2003-08-07  MOMOSE Tsuyoshi  <t-momose@netlab.nec.co.jp>
	* kame/kame/had/mpa.c,haadisc.c
	A home address should be passed in the source address of a mobile
	solicit pakcet. This problem was reported and donated the code by
	Kenichi Yajima <yajima@netlab.nec.co.jp>
	
2003-08-07  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6/cftoken.l: use a stronger lexical check for
	interface names.
	From: Hajimu UMEMOTO <ume@mahoroba.org>
	* kame/kame/dhcp6/config.c (configure_interface): check interfaces
	specified in the configuration file really exist.

2003-08-05  Hideki ONO  <ono@soft.net.fujitsu.co.jp>
	* removed vrrp6 support because it might be a patent infringement.
	  http://www.ietf.org/ietf/IPR/cisco-ipr-draft-ietf-vrrp-ipv6-spec.txt

2003-08-01  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6: implemented various clarifications and some new
	  features:
	- supported DNS name and NTP servers options
	- dhcp6s now uses the server/relay port to send a relay-rely
	  message
	- dhcp6s now ignores a rebind message when it cannot find a
	  binding, according to prefix-delegation-04
	- dhcp6relay now accepts relay-reply messages on the socket
	  listening on the server/relay port
	* The DHCPv6 implementation is now built by default under
	  *bsd/usr.sbin/.

<200307>
Thu Jul 31 19:19:48 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_mncore.c,mip6_cncore.c
	sending/receiving a binding refresh request message is supported.

Tue Jul 29 17:09:15 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_mncore.[hc]
	select a CoA using the source address selection like algorithm.
	suggested by Francis.Dupont@enst-bretagne.fr.

2003-07-28  MOMOSE Tsuyoshi  <t-momose@netlab.nec.co.jp>
	* sys/netinet6/{ip6_output.c, ip6_var.h, mip6_cncore.c,
	mip6_hacore.[ch], nd6.c}:
	CN or HA dosen't add a type2 routing header to a Binding
	Acknowledgement packet when a deregistration packet was issued
	on home link and it was failed by some reasons.
	reported by Yukiyo.Akisada@jp.yokogawa.com and v6pc cert.
	team.

Mon Jul 28 20:03:46 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_cncore.c,mip6_mncore.[hc],in6.c,nd6_rtr.c
	  kame/sys/net/if_hif.[hc]
	assign a CoA separately for each home interface, since a home
	interface is logically independent from each other.

Mon Jul 28 14:38:11 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_hacore.c,mip6_cncore.c:
	- fixed a bug not sending a binding ack when DAD for a link-local
	  address is failed.  (en-bugged during re-structuring BC logic.)
	- fixed a bug using RTHDR2 with Mobility Headers other than
	  a binding ack.
	reported by Yukiyo.Akisada@jp.yokogawa.com.

Fri Jul 25 19:12:15 JST 2003  itojun@iijlab.net
	* sys/netinet/ah_core.c: hmac-ripemd160 support

Fri Jul 25 18:07:43 JST 2003 sakane@kame.net
	* kame/sys/netkey/key.c
	fixed that the kernel crashed when key_spdacquire() was called
	because key_spdacquire() had been implemented imcopletely.

Thu Jul 24 16:11:06 JST 2003	keiichi@iij.ad.jp
	* mip6 related files.
	simplify the prefix and advertising router list management
	mechanism on a mobile node.

2003-07-20  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6/config.c (configure_duid): fixed a bug of
	allocating short memory.  Upgrading is required if you configure
	dhcp6s with the duid statement.

Sat Jul 19 18:12:01 JST 2003  itojun@iijlab.net
	* sys/netinet6/esp_aesctr.c: support draft-ietf-ipsec-ciph-aes-ctr-03
	  as there's no official DOI assignment yet, we use private DOI number.
	* sys/netinet6/ah_aesxcbcmac.c: support
	  draft-ietf-ipsec-ciph-aes-xcbc-mac-03.
	  as there's no official DOI assignment yet, we use private DOI number.

Tue Jul 15 20:02:05 JST 2003  itojun@iijlab.net
	* sys/netinet6/esp_rijndael.c: simplify and update rijndael code.
	  markus@openbsd

2003-07-14  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6: revised the relay agent implementation
	based on dhcpv6-28.
	- dhcp6relay now relays packets between clients and servers using
	  relay forward/reply messages.
	- dhcp6s now accepts relay forward messages and replies with relay
	  reply messages.

Sat Jul 12 15:05:23 JST 2003	suz@crl.hitachi.co.jp
	* openbsd/sys/netinet/igmp_var.h
	makes IGMPV3-kernel compilable on openbsd.

Thu Jul 10 21:35:04 JST 2003	keiichi@iij.ad.jp
	* kame/kame/had/halist.c
	fixed a bug that a DHAAD reply message only includes one global
	address per home agent.  the message must include all global
	addresses of all home agents.
	reported by Yukiyo.Akisada@jp.yokogawa.com.

2003/07/10 21:07:50 JST kjc@csl.sony.co.jp
	the first step for transition to pf/altq.
	marge pf/altq into the existing altq-3.
	allow both pf/altq and altq-3 to coexit.
	currently, ALTQ3_COMPAT and ALTQ3_CLFIER_COMPAT are defined
	in sys/altq/altq.h to support altq-3.
	altq-3 will remain in kame for research experiments but will
	not be merged into bsd releases.

2003-07-08  Hideki ONO  <ono@soft.net.fujitsu.co.jp>
	* kame/sys/netinet6/{in6_ifattach.c,ip6_mroute.c,ip6_mroute.h}
	protect against interface removal.

Tue Jul  8 18:51:06 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_mncore.c
	insert alternate careof address sub-option when creating a
	binding update for the home registration.  This sub-option is
	MUST when we protect the message with ESP and strictly speaking,

kame/kame/pvcsif/pvcbridge.c
/*
 * Copyright (C) 1999
 *	Sony Computer Science Laboratories, Inc.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY SONY CSL AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL SONY CSL OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 * $Id: pvcbridge.c,v 1.1 2000/01/19 12:47:08 kjc Exp $
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <ctype.h>
#include <err.h>

#include <sys/socket.h>
#include <sys/ioctl.h>
#include <sys/ioctl.h>
#include <net/if.h>

#include <net/if_atm.h>

enum bridge_op { DELETE, ADD, GET };
static void usage(void);
static int pcr2str(int pcr, char *buf);

static void 
usage(void)
{
	fprintf(stderr, "usage: pvcbridge [add|delete] pvc_if1 pvc_if2\n");
	fprintf(stderr, "       pvcbridge [get] pvc_if\n");
	exit(1);
}

int verbose = 1;
int operation = GET;

int 
main(int argc, char **argv)
{
	struct pvcfwdreq pvcreq;
	struct pvctxreq pvctxreq, pvctxreq2;
	char *if_name, *if_name2;
	char speed[64], speed2[64];
	int s;

	if (argc < 2)
		usage();

	argc--; argv++;
	if (strcmp(*argv, "add") == 0) {
		operation = ADD;
		argc--; argv++;
	}
	else if (strcmp(*argv, "delete") == 0) {
		operation = DELETE;
		argc--; argv++;
	}
	if (strcmp(*argv, "get") == 0) {
		operation = GET;
		argc--; argv++;
	}

	if (argc > 0) {
		if_name = *argv;
		argc--; argv++;

		if (strncmp(if_name, "pvc", 3) != 0)
			usage();
			
		if (operation != GET && argc > 0) {
			if_name2 = *argv;
			argc--; argv++;

			if (strncmp(if_name2, "pvc", 3) != 0)
				usage();
		}
	}
	else
		usage();

	if ((s = socket(AF_INET, SOCK_DGRAM, 0)) < 0)
		err(1, "can't open socket");

	pvcreq.pvc_ifname[IFNAMSIZ-1] = '\0';
	strncpy(pvcreq.pvc_ifname, if_name, IFNAMSIZ-1);

	if (operation != GET) {
		pvcreq.pvc_ifname2[IFNAMSIZ-1] = '\0';
		strncpy(pvcreq.pvc_ifname2, if_name2, IFNAMSIZ-1);
		pvcreq.pvc_op = operation;

		if (ioctl(s, SIOCSPVCFWD, &pvcreq) < 0)
			err(1, "SIOCSPVCFWD");

		if (operation == DELETE) {
			printf("pvc bridging %s <--> %s deleted\n",
			       pvcreq.pvc_ifname, pvcreq.pvc_ifname2);
			return (0);
		}
	}

	if (ioctl(s, SIOCGPVCFWD, &pvcreq) < 0)
		err(1, "SIOCSPVCFWD");

	/*
	 * print info
	 */
	if (pvcreq.pvc_ifname2[0] == '\0') {
		printf("pvc bridging is not set on %s\n",
		       pvcreq.pvc_ifname);
		return (0);
	}

	/*
	 * get tx info
	 */
	bzero(&pvctxreq, sizeof(pvctxreq));
	strncpy(pvctxreq.pvc_ifname, pvcreq.pvc_ifname, IFNAMSIZ-1);
	if (ioctl(s, SIOCGPVCTX, &pvctxreq) < 0)
		err(1, "SIOCSPVCTX");
	bzero(&pvctxreq2, sizeof(pvctxreq2));
	strncpy(pvctxreq2.pvc_ifname, pvcreq.pvc_ifname2, IFNAMSIZ-1);
	if (ioctl(s, SIOCGPVCTX, &pvctxreq2) < 0)
		err(1, "SIOCSPVCTX");

	pcr2str(pvctxreq.pvc_pcr, speed);
	pcr2str(pvctxreq2.pvc_pcr, speed2);
	printf("pvc bridging %s:[%d:%d](%s) <--> %s:[%d:%d](%s)\n",
	       pvcreq.pvc_ifname,
	       ATM_PH_VPI(&pvctxreq.pvc_aph),
	       ATM_PH_VCI(&pvctxreq.pvc_aph),
	       speed,
	       pvcreq.pvc_ifname2, 
	       ATM_PH_VPI(&pvctxreq2.pvc_aph),
	       ATM_PH_VCI(&pvctxreq2.pvc_aph),
	       speed2);
	close(s);

	return (0);
}

static int 
pcr2str(int pcr, char *buf)
{
	if (pcr < 0)
		sprintf(buf, "invalid\n");
	else if (pcr == 0)
		sprintf(buf, "full speed");
	else if (pcr < 1000)
		sprintf(buf, "%dbps", pcr * 48 * 8);
	else if (pcr < 1000000)
		sprintf(buf, "%.2fKbps", (double)pcr * 48 * 8 / 1000);
	else
		sprintf(buf, "%.2fMbps", (double)pcr * 48 * 8 / 1000000);
	return (strlen(buf));
}
